src/Infrastructure.EntityFramework/Tools/Repositories/SendRepository.cs

```diff
@@ -69,4 +70,28 @@ public async Task<ICollection<Core.Tools.Entities.Send>> GetManyByUserIdAsync(Gu
             return Mapper.Map<List<Core.Tools.Entities.Send>>(results);
         }
     }
+
+    public UpdateEncryptedDataForKeyRotation UpdateForKeyRotation(Guid userId,
+        IEnumerable<Core.Tools.Entities.Send> sends)
+    {
+        return async (_, _) =>
+        {
+            var newSends = sends.ToList();
+            using var scope = ServiceScopeFactory.CreateScope();
+            var dbContext = GetDatabaseContext(scope);
+            var userSends = await GetDbSet(dbContext)
+                .Where(s => s.UserId == userId)
+                .ToListAsync();
+            var validSends = userSends
+                .Where(send => newSends.Any(newSend => newSend.Id == send.Id));
+            foreach (var send in validSends)
+            {
+                var updateSend = newSends.First(newSend => newSend.Id == send.Id);
+                send.Key = updateSend.Key;
+            }
```

Comment:
🎨 If you turn newSends into a dictionary, this would be a key lookup instead of a linear search.

```cs
var newSends = sends.ToDictionary(s => s.Id);

// ... in foreach
send.Key = newSends[send.Id].Key
```

Reply:
Good catch!